core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/docker/DockerContainer.scala

```diff
@@ -281,9 +309,11 @@ class DockerContainer(protected val id: ContainerId,
       }
       .via(new CompleteAfterOccurrences(_.containsSlice(DockerContainer.byteStringSentinel), 2, waitForSentinel))
       // As we're reading the logs after the activation has finished the invariant is that all loglines are already
-      // written and we mostly await them being flushed by the docker daemon. Therefore we can timeout based on the time
+      // written and we mostly await them being flushed by the docker daemon. Therefore we can time out based on the time
       // between two loglines appear without relying on the log frequency in the action itself.
       .idleTimeout(waitForLogs)
+      // Apply an overall time limit for this log collecting and processing stream.
+      .completionTimeout(logProcessingTimeout)
```

Comment:
It looks to me like this is the crux of the patch - limit the log extraction until sentinel is found. I don’t know what a reasonable rate should be but up to 2s/1MB strikes me as long. I like Markus’ suggestion and wonder what the drain rate might be under load. That’s especially when waiting is on the critical path.

Reply:
It's not really the crux but an additional safeguard as I read it. It's a "catchall".

The crux here is to ignore sentinels (and thus don't tail logs, which guarantees a finish of the stream) in error cases.

Reply:
See my comment above on why I thought/think this is the essence of the patch. Arguably this limit should have already existed when waiting for logs to drain. But I understand the combined effect as well now, thanks to the excellent comments.